pyccel/codegen/printing/fcode.py

```diff
@@ -730,7 +730,25 @@ def _print_MathFloor(self, expr):
 
     def _print_NumpyFloor(self, expr):
         result_code = self._print_MathFloor(expr)
-        return 'real({})'.format(result_code)
+        if isinstance(expr.dtype, NativeInteger):
```

Comment:
numpy floor is always a real no ?, and also it doesn't take a complex

Reply:
Like every universal Numpy function, it takes a `dtype` optional argument: https://numpy.org/doc/stable/reference/ufuncs.html#optional-keyword-arguments

Reply:
https://numpy.org/doc/stable/reference/generated/numpy.floor.html

No. Most of numpy's functions have an optional `dtype` kwarg. If this is not specified then numpy floor will return a real. However if it is specified then it will return whatever dtype is given. This dtype is unrelated to the arguments. Hence:
```
> import numpy as np
> np.floor(3.5, dtype=complex)
```
should return a complex(3.0, 0). It looks like there's a problem with this at the moment though https://stackoverflow.com/questions/47838306/getting-no-loop-matching-the-specified-signature-and-casting-error
I can't replicate the documented behaviour on my computer.

My reason for implementing this was simply so that I can get a randint using numpy floor to get the int, rather than using math floor which does not work for arrays. As it was it casts the result to a real and I would then need to cast back to an int

Reply:
you mean that we can do something like `np.floor(2.3, dtype='int')` ?

Reply:
In numpy1.18 it gives me an error

Reply:
Sorry, you're right my link wasn't much use
http://numpy-discussion.10968.n7.nabble.com/floor-with-dtype-td44734.html
It seems that the dtype refers to the input and the output argument for some reason.

Can you think of another way to represent this for an array without double casting? Or do we need to have 2 implementations of floor, one for numpy's floor and one for randint?

Reply:
yes maybe we can use a floor for randint  alone

Reply:
So here is the situation:

> dtype
> >    New in version 1.6.
> >    Overrides the dtype of the calculation and output arrays. Similar to signature.

> signature
> >   Either a data-type, a tuple of data-types, or a special signature string indicating the input and output types of a ufunc. This argument allows you to provide a specific signature for the 1-d loop to use in the underlying calculation. If the loop specified does not exist for the ufunc, then a TypeError is raised. Normally, a suitable loop is found automatically by comparing the input types with what is available and searching for a loop with data-types to which all inputs can be cast safely. This keyword argument lets you bypass that search and choose a particular loop. A list of available signatures is provided by the types attribute of the ufunc object. For backwards compatibility this argument can also be provided as sig, although the long form is preferred. Note that this should not be confused with the generalized ufunc signature that is stored in the signature attribute of the of the ufunc object.

So `dtype` and `signature` allow to change the type of the input and output arrays, if the required type is actually allowed.